### ISSUE TITLE
Why does dreamer not require stored hidden states?

### ISSUE THREAD

**Issue by xlnwel** [danijar/dreamer#24]
Hi, @danijar 

I notice that Dreamer does not store the hidden states of RSSM in the replay and always uses the zero initialization during learning. This seems to work reasonably well with Dreamer, but not with other off-policy algorithms such as R2D2. I've tried to add hidden states and burn-in periods to the replay but did not spot any performance difference. Do you have any idea why this is the case?

By the way, could you please publish the hyperparameters you use in Atari games?

**Comment by E666GT** [comment]
Maybe in the dreaming behavior learning, it uses online learning？

**Comment by xlnwel** [comment]
Hi @E666GT,

I humbly don't think so. It does not seem to be related to on-policy or off-policy learning. PPO is on-policy but if we always use zero initialization for the LSTM state, it'll mess up.

**Comment by danijar** [comment]
Can you point me to evidence that PPO messes up when trained on sequences where the initial states is zero? R2D2 also works quite well with zero state, although it works better with stale remembered state.

**Comment by xlnwel** [comment]
Hi @danijar 

I re-ran several experiments yesterday and found that PPO without stored states performed worse than PPO with stored states and in some environments the former did not learn at all. The following figure shows PPO on [Procgen CoinRun](https://github.com/openai/procgen)

<img width="339" alt="image" src="https://user-images.githubusercontent.com/10184153/85823065-fe7cb680-b7ae-11ea-9db8-402b50891f5d.png">

where the gray curve is PPO with stored states and the orange curve is PPO without stored states. We can see that PPO without stored states is more unstable. Although this could be mitigated by lowering the learning rate(the blue curve is PPO without stored states using `lr=1e-4`, whereas the others use `lr=5e-4`), PPO with stored states is still superior.

The following figure shows another experiments on PongNoFrameskip-v0

<img width="321" alt="image" src="https://user-images.githubusercontent.com/10184153/85822863-81e9d800-b7ae-11ea-999e-b21d4001b582.png">

where the red curve is PPO with stored states and the blue curve is PPO without stored states. In this experiment, the latter fails to learn anything. 

Due to resource limits, I did not do thorough experiments. Maybe there are some configurations that work for PPO without stored states, but according to my experience, PPO with stored states is better.

**Comment by danijar** [comment]
Cool. Did you use burn in, where you ignore losses for the first few time steps on the training sequence?

Storing recurrent states for PPO is easy because it's on policy. For off-policy agents like Dreamer and R2D2 it becomes more complicated because stored states will become stale.

One reason the RSSM world model might be more robust than a plan LSTM is it's stochasticity that causes the model to learn how to deal with a higher diversity of states during training.

**Comment by xlnwel** [comment]
Hi @danijar 

Yeah, I've tried burn-in, but no big difference was noticed on environments like Walker Run and Finger Spin from DMC.

Thanks for the explanation. I think it's reasonable to think it in that way.

I cannot run more experiments on DMC as my Mujoco license is expired. Could you please publish the hyperparameters used on Atari games. That'll save me a lot of time. Thanks in advance!

**Comment by danijar** [comment]
Yes, I'm working on an update and will push it once it's ready.
